File: LeetCode/Java/142-two-pointers-环形链表-ii.java
/*
 * @lc app=leetcode.cn id=142 lang=java
 *  [142] 环形链表 II
 */

// @lc code=start

class Solution {
    public ListNode detectCycle(ListNode head) {
        var visited = new HashSet<ListNode>();

        ListNode node = head;
        while (node != null) {
            if (visited.contains(node))
                return node;
            visited.add(node);
            node = node.next;
        }

        return null;
    }
}
// @lc code=end

File: LeetCode/Go/142-two-pointers-环形链表-ii.go
/*
 * @lc app=leetcode.cn id=142 lang=golang
 *
 * [142] 环形链表 II
 */
package leetcode

// @lc code=start
func detectCycle(head *ListNode) *ListNode {
	var p = head
	var set = map[*ListNode]bool{}
	for p != nil {
		if _, ok := set[p]; ok {
			return p
		}
		set[p] = true
		p = p.Next
	}
	return nil
}

// @lc code=end
